File: api/src/user/user.controller.ts
import { Controller, Get, Patch, Param, Query, Body, UseGuards } from '@nestjs/common';
import { UserService } from './user.service';
import { JwtAuthGuard } from '../common/guards/jwt-auth.guard';
import { CurrentUser } from '../common/decorators/current-user.decorator';
import { UpdateUserDto } from './dto/update-user.dto';
import { SearchUserDto } from './dto/search-user.dto';

@Controller('users')
@UseGuards(JwtAuthGuard)
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get('me')
  async getMe(@CurrentUser() user: { id: string }) {
    return this.userService.findById(user.id);
  }

  @Get('search')
  async searchUsers(@Query() dto: SearchUserDto) {
    return this.userService.searchByNickname(dto.nickname);
  }

  @Get(':id')
  async getUser(@Param('id') id: string) {
    return this.userService.findById(id);
  }

  @Patch(':id')
  async updateUser(@Param('id') id: string, @Body() dto: UpdateUserDto) {
    return this.userService.update(id, dto);
  }
}

File: api/src/main/java/com/chatops/domain/user/controller/UserController.java
package com.chatops.domain.user.controller;

import com.chatops.domain.auth.dto.UserResponse;
import com.chatops.domain.user.dto.UpdateUserRequest;
import com.chatops.domain.user.entity.User;
import com.chatops.domain.user.service.UserService;
import com.chatops.global.redis.RedisService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Size;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Map;

@Validated
@RestController
@RequestMapping("/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;
    private final RedisService redisService;

    @GetMapping("/me")
    public UserResponse getMe(@AuthenticationPrincipal User user) {
        return userService.findById(user.getId());
    }

    @GetMapping("/search")
    public List<UserResponse> searchUsers(@RequestParam @Size(min = 2, max = 20) String nickname) {
        return userService.searchByNickname(nickname);
    }

    @GetMapping("/{id}")
    public UserResponse getUser(@PathVariable String id) {
        return userService.findById(id);
    }

    @PatchMapping("/{id}")
    public UserResponse updateUser(@AuthenticationPrincipal User user,
                                   @PathVariable String id,
                                   @Valid @RequestBody UpdateUserRequest request) {
        if (!user.getId().equals(id)) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Cannot update another user's profile");
        }
        return userService.update(id, request);
    }

    @GetMapping("/online-status")
    public Map<String, Boolean> getOnlineStatus(@RequestParam @Size(max = 100) List<String> userIds) {
        return redisService.getOnlineStatuses(userIds);
    }
}
